tests/solaar/ui/test_pair_window.py

```diff
@@ -0,0 +1,128 @@
+
+    assert receiver.pairing.lock_open == lock_open
+    assert receiver.pairing.discovering == discovering
+
+
+@pytest.mark.parametrize(
+    "receiver, expected_result, expected_error",
+    [
+        (TestReceiver("unifying", "unifying", True), True, False),
+        (TestReceiver("unifying", "unifying", False), False, True),
+        (TestReceiver("bolt", "bolt", True), True, False),
+        (TestReceiver("bolt", "bolt", False), False, True),
+    ],
+)
+def test_prepare(receiver, expected_result, expected_error):
+    result = pair_window.prepare(receiver)
+
+    assert result == expected_result
+    assert bool(receiver.pairing.error) == expected_error
+
+
+@pytest.mark.parametrize(
+    "receiver, lock, count, result",
+    [
+        (TestReceiver("fail", "unifying", False, receiver.Pairing(lock_open=False)), False, 2, False),
+        (TestReceiver("succeed", "unifying", True, receiver.Pairing(lock_open=True)), True, 1, True),
+        (TestReceiver("error", "unifying", True, receiver.Pairing(error="error")), False, 0, False),
+        (TestReceiver("new device", "unifying", True, receiver.Pairing(new_device=TestDevice())), False, 2, False),
+        (TestReceiver("closed", "unifying", True, receiver.Pairing()), False, 2, False),
+        (TestReceiver("closed", "unifying", True, receiver.Pairing()), False, 1, False),
+        (TestReceiver("closed", "unifying", True, receiver.Pairing()), False, 0, False),
+        (TestReceiver("fail bolt", "bolt", False), False, 1, False),
+        (TestReceiver("succeed bolt", "bolt", True), True, 0, True),
+        (TestReceiver("error bolt", "bolt", True, receiver.Pairing(error="error")), False, 2, False),
+        (TestReceiver("new device", "bolt", True, receiver.Pairing(new_device=TestDevice())), True, 1, False),
+        (TestReceiver("discovering", "bolt", True, receiver.Pairing()), True, 1, True),
+        (TestReceiver("closed", "bolt", True, receiver.Pairing()), False, 2, False),
+        (TestReceiver("closed", "bolt", True, receiver.Pairing()), False, 1, False),
+        (TestReceiver("closed", "bolt", True, receiver.Pairing()), False, 0, False),
+        (
+            TestReceiver("passkey", "bolt", True, receiver.Pairing(device_passkey=50, device_authentication=0x01)),
+            True,
+            0,
+            True,
+        ),
+        (
+            TestReceiver("passkey", "bolt", True, receiver.Pairing(device_passkey=50, device_authentication=0x02)),
+            True,
+            0,
+            True,
+        ),
+        (TestReceiver("add", "bolt", True, receiver.Pairing(device_address=2, device_name=5), pairable=True), True, 2, True),
+        (TestReceiver("add", "bolt", True, receiver.Pairing(device_address=2, device_name=5), pairable=False), True, 2, False),
+        (TestReceiver("add fail", "bolt", False, receiver.Pairing(device_address=2, device_passkey=5)), False, 2, False),
+    ],
+)
+def test_check_lock_state(receiver, lock, count, result):
```

Comment:
You are testing quite many different scenarios here. If you'd have a third function in this test, then it would probably grow to double the size or more ;)

However, there's a solution for this when code is written with testability in mind. Therefore, the _create_ and _check_lock_state functions need to be unit testable with like 5 lines of code and parametrized input to cover possible inputs. The same for create which also has some branches  in it.

Then in this test, that involves multiple functions and is like an integration tests, you only test one or two variations. The _create_ and _check_lock_state functions behave as expected and ensured by the unit tests and an integration test, then checks, that the composition of them also works. For that, even testing a single path is fine.